### PR TITLE
DEV: Set limits for text columns in the DB

### PR DESCRIPTION
**Pull request by Flink**

### CODE REVIEW

**Comment by jjaffeux** [comment]
@danielwaterworth would you have the time to look at this one please?

**Comment by davidtaylorhq** [comment]
What happens if an existing value exceeds the new maximum length? Is this migration going to blow up? 🤔  (should we truncate instead??)

Also, given the number of changes here, perhaps we need to skip the migration transaction?

**Comment by Flink** [comment]
> What happens if an existing value exceeds the new maximum length? Is this migration going to blow up?  (should we truncate instead??)

Yes the migration blows up when that happens but since it’s in a transaction, everything is reverted. Truncating would mean have potentially invalid data, so not sure about that :thinking: 

> Also, given the number of changes here, perhaps we need to skip the migration transaction?
I’m not sure how much time this will take to complete, but `TEXT` and `VARCHAR` in PG are virtually the same thing so I guess not that much? We could try to run it on a rather large dataset to check though.

**Comment by davidtaylorhq** [comment]
> Yes the migration blows up

In that case I think we should try to make it print a nice error message with a link to instructions on Meta for self-hosters to identify and resolve the issues. Otherwise, I think we're likely to see a number of support topics.

**Comment by Flink** [comment]
> In that case I think we should try to make it print a nice error message with a link to instructions on Meta for self-hosters to identify and resolve the issues.

I added a message that is displayed as bold and red, what do you think? (feel free to improve the message too :grin:)
https://github.com/discourse/discourse/pull/20957/files#diff-a65e4dabc5fd58fe5f7ced915870c77cf7c95374afaa7559cf165f0988dc660aR81

**Comment by danielwaterworth** [comment]
> I’m not sure how much time this will take to complete, but `TEXT` and `VARCHAR` in PG are virtually the same thing so I guess not that much? We could try to run it on a rather large dataset to check though.

Doesn't it need to scan the whole table to see if there are entries that are too large?

**Comment by Flink** [comment]
> Doesn't it need to scan the whole table to see if there are entries that are too large?

I don’t really know, but yes, what you’re saying makes sense.

File: lib/validators/stripped_length_validator.rb [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 class StrippedLengthValidator < ActiveModel::EachValidator
   def self.validate(record, attribute, value, range)
-    if value.nil?
+    if value.blank?
```
Comment (ZogStriP) [inline lib/validators/stripped_length_validator.rb:5]: Should this be tested? Should also probably be in another PR so we keep that PR solely on DB limit changes?

**Comment by ZogStriP** [comment]
> I’m not sure how much time this will take to complete, but `TEXT` and `VARCHAR` in PG are virtually the same thing so I guess not that much? We could try to run it on a rather large dataset to check though.

Please definitely do that before merging and deploying this PR 👍

File: lib/validators/stripped_length_validator.rb (hunk 1 above)
Comment (Flink) [reply lib/validators/stripped_length_validator.rb:5]: Actually this is tested by the shoulda matcher we already have. It seems changing the column type from `text` to `string` makes it test `""` as a value instead of just `nil`. (that’s how I caught this bug in the first place :sweat_smile:)
